contributing/guide-observability.md

```diff
@@ -120,6 +120,61 @@ logs don't show up in production normally.
 For the question of what constitutes an error, errors should be actionable by a
 human. See the [Dave Cheney article] on this topic for some more discussion.
 
+## Good Errors
+
+An error message is good if it communicates
+
+1. what went wrong
+2. where it went wrong
+3. for which inputs
+
+and if it is concise without redundancies and at best displayable in one line.
+
+The general goal of an error message is to be actionable for the user who is trying
+to avoid the error in the future.
+
+Examples for bad error messages:
+- `cannot resolve package dependencies: Invalid Semantic Version`
+- `cannot parse file foo.yaml: failed to parse YAML of file foo.yaml`
+
+Examples for how to improve them:
+- `cannot resolve package dependencies in crossplane.yaml: invalid Semantic Version "main" on package "provider-kubernetes"`
+- `failed to load OCI image provider-gcp/v0.42.0: failed to unpack layer: cannot parse file: failed to parse YAML of file foo.yaml: unexpected indention in line 4`.
+
+General rules:
+1. follow the pattern `generic error cause: more details: even more details`.
+2. put values into quotes (`%q` in Golang), with the exception of
+   - resource names with namespace or resource type (e.g. `default/foo` or `providers/foo`), resource names alone are quoted
+   - kinds (e.g. `cannot find provider "provider-kubernetes"`; here, the kind provider is without quotes, the provider name is with)
+   - filenames with extensions and/or path (e.g. `cannot parse foo.yaml`).
+3. add a context of e.g. filenames, object names, or values which are wrong.
+
+   Which context is added by which functions depends on the level of abstraction of the code at hand: the package loader will know the OCI image name, the YAML parser might only see a byte slice. The former adds the image name as context, some layer in-between adds the filename and the YAML parser adds the line and column in the byte stream.
+
+   Filenames should be relative to the logical working directory of the context at hand (e.g. a package root, current working directory, workspace root). Ensure that the context is clear in the error message (e.g. `failed to load package provider-kubernetes: failed to parse apis/xrd.yaml`).
+
+6. add a context just once.
+
+   Rule of thumb: if a function gets a parameter, it will also include it in the error messages, so the caller does not have to.
+   
+   E.g. a function getting a filename as input will include the filename in errors it returns.
+7. don't return multi-line errors.
+8. aggregate errors if there is clear use for the user, but not in general, i.e. fail fast is the default.
+
+   Aggregate through `kerrors.NewAggregate`. When the number of aggregated errors is unbounded,
+   return a subset, e.g. the first 3.
+
+   E.g.: `cannot parse workspace: failed to parse foo.yaml, bar.yaml, abc.yaml, and 17 more files`
+9. don't include sensitive information like tokens into error messages, not even in debug mode.
```

Comment:
Given that the log line author may not necessarily know what is being placed in the field value and whether or not it's sensitive, I wonder if we can cover this by asserting that field/property names are fine to include, but there isn't a reason to log the value of the field.

For example:
```
something went wrong: invalid type for field spec.forProvider.tag.owner: expected bool, got string
```

versus:
```
something went wrong: invalid type for field spec.forProvider.tag.owner: expected bool, got string, value: user@example.com
```

The first example still gives the author/user enough information to track down the problem without potentially leaking sensitive information, whereas the second leaks sensitive information.

This also looks like it may be in direct opposition to part of (3) above:
>add a context of e.g. filenames, object names, or values which are wrong.

specifically, "or values which are wrong".

Reply:
If you know the field is not sensitive (like the `type` example, enum style), you can log it or include it into an error. If you write generic code (against any kind of yaml), be more careful. Field names can always be logged.

I guess a username is somewhere in a grey area, maybe depending on legislation.

A common example are object names. They are never encrypted in kube, not even with encryption at REST aka encryption in etcd. Of course you can store your password in the name of an object. But 🤷‍♂️ 

So as rule, maybe "log it if you know it is not sensitive, e.g. for enum fields, object names, file names".

Reply:
@tnthornton updated. Please double check.

Reply:
>or values which are wrong (and the values are known to be insensitive)

I think the clarification works for me 👍.

Small nit - Your summation above "log it if you know it is not sensitive, e.g. for enum fields, object names, file names, otherwise don't" seems more clear to me versus strictly just "values which are wrong (and the values are known to be insensitive)". However, this is probably more subjective and I don't think it should block on the current progress.

Thanks @sttts !